### PR TITLE
Add uptime monitoring to riemann-health

### PR DESCRIPTION
**Pull request by smortex**
Make a new metric with the node uptime available.  The metric is the raw uptime in seconds and the description is a human readable duration.  Critical and Warning thresholds are reversed when compared to other metrics: a value lower that these thresholds indicates a problem. This is intended to get notifications when nodes restart unexpectedly.

The uptime check is not enabled by default and must be explicitly enabled using the `--checks` flag.

### CODE REVIEW

File: lib/riemann/tools/health.rb [hunk 1]
```diff
@@ -240,47 +270,50 @@ def sunos_memory
...
+
         raw = `top -l 1 | grep -i "^\\(cpu\\|physmem\\|load\\)"`.chomp
-        @topdata = { stamp: Time.now.to_i }
+        topdata = {}
         raw.each_line do |ln|
           if ln.match(/Load Avg: [0-9.]+, [0-9.]+, ([0-9.])+/i)
-            @topdata[:load] = Regexp.last_match(1).to_f
+            topdata[:load] = Regexp.last_match(1).to_f
           elsif ln.match(/CPU usage: [0-9.]+% user, [0-9.]+% sys, ([0-9.]+)% idle/i)
-            @topdata[:cpu] = 1 - (Regexp.last_match(1).to_f / 100)
```
Comment (jamtur01) [inline lib/riemann/tools/health.rb:249]: Are these unrelated stylistic changes?

**Comment by jamtur01** [comment]
You thinking about replacing some of the more complex regex's etc with racc grammars?

**Comment by smortex** [comment]
This is admittedly oven-engineered but the `uptime(1)` commands try hard to [format the uptime for humans](https://cgit.freebsd.org/src/tree/usr.bin/w/w.c#n513) and a parser is probably the most maintainable (but not straightforward) way to understand things intended to be read by humans…

That might be a direction we are not ready to take, but I would benefit from this metric and would like to avoid to spit it apart, hence the PR for a discussion :smile:

**Comment by jamtur01** [comment]
I'm good with it. I am curious if there are other things we could do easier with it.

File: lib/riemann/tools/health.rb (hunk 1 above)
Comment (smortex) [reply lib/riemann/tools/health.rb:249]: I split this in 3 commits. This is part of the first which is expected to not change the behavior of the program but makes data-caching more abstract: instead of just having cache for `top(1)` on MacOS, we have a "global" cache that just cache the output of top on MacOS.

The other commits then use this to also cache the output of `uptime` and the parser result.

Comment (jamtur01) [reply lib/riemann/tools/health.rb:249]: Cool. 👍🏻

**Comment by smortex** [comment]
For now, it's the only time I see something that is likely to break easily is we rely on regexps. The previous code that was loading the 1 minute load average did it in a way that worked almost everywhere: "first thing that looks-like a float after the last `:`", here are examples from the test suite I collected in the real world today:

* `11:10  up  3:40, 1 user, load averages: 0,25 0,67 0,68`
* `11:30AM  up 4 hrs, 1 user, load averages: 0.45, 0.53, 0.54`
*  `11:46  up 38 days, 22:21, 2 users, load averages: 1,76 1,24 0,94`
* ` 10:40:21 up 1 day, 18:51,  1 user,  load average: 0,46, 1,45, 2,00`
* ` 11:50:17 up 1 day, 20:01,  1 user,  load average: 1.66, 1.69, 1.38`

only when the locale implied a `,` as decimal separator we lost the fractional part, not really an issue because the code would run as a service with the default locale I guess, so this was okayish.

Extracting the uptime from this mess with regexp would really be another story.

All other regexp I see seem to extract data from well defined strings, the output of top on MacOS might be considered fragile but the regexps are straightforward so unless they update the output at some point I think this is fine as it is now.

**Comment by smortex** [comment]
I could not test that the first commit does not change the behavior on MacOS, so if you can confirm that noting breaks it would be awesome.  I spotted a minor issue in the first commit and fixed it, I consider this ready.

**Review by jamtur01 — APPROVED** [review]
LGTM
